reana_client/api/client.py

```diff
@@ -290,21 +294,24 @@ def create_workflow_from_json(
         )
     try:
         reana_yaml = dict(workflow={})
-        if workflow_file:
-            reana_yaml["workflow"]["file"] = workflow_file
-        else:
-            reana_yaml["workflow"]["specification"] = workflow_json
         reana_yaml["workflow"]["type"] = workflow_engine
         if parameters:
             reana_yaml["inputs"] = parameters
         if outputs:
             reana_yaml["outputs"] = outputs
+        if workflow_file:
+            reana_yaml["workflow"]["file"] = workflow_file
+            reana_yaml["workflow"][
+                "specification"
+            ] = load_workflow_spec_from_reana_yaml(reana_yaml)
+            input_params = load_input_parameters(reana_yaml)
+            if input_params:
+                reana_yaml["inputs"]["parameters"] = input_params
+        else:
+            reana_yaml["workflow"]["specification"] = workflow_json
         validate_reana_yaml(reana_yaml)
-        reana_specification = reana_yaml
         (response, http_response) = current_rs_api_client.api.create_workflow(
-            reana_specification=json.loads(
-                json.dumps(reana_specification, sort_keys=True)
-            ),
+            reana_specification=json.loads(json.dumps(reana_yaml, sort_keys=True)),
```

Comment:
That's a good question. Passing directly `reana_yaml` appears to be working, but I am a bit hesitant to remove the whole `json.loads(json.dumps(...))` thing. 

A not-so-likely guess in my opinion, is that it is needed to sort the keys of the dictionary (note the `sort_keys=True`): starting from Python 3.7, indeed, [the dictionary order is guaranteed to be insertion order](https://docs.python.org/3.7/library/stdtypes.html#dict). However, I don't see why it would be needed, if not only for debugging purposes.

My most educated guess would instead be that it is needed to ensure that the `reana_specification` parameter is JSON-serializable (especially because this is a r-server API call), or anyway in a JSON-compatible format (e.g. by replacing tuples with lists). Since the dictionary is built using user-provided arguments, `reana_yaml` could in theory contain non-serializable objects: in that case, `json.dumps()` will raise an exception.

Reply:
Let's keep it like this, I could not find any explanation of why it's done like this in older commits.